Disable send button if the user has no apps

In the push message tab, on a new gotify server, that has no app set up. When you press the push message, with other fields filled,The app Freezes.

![Screenshot_2021-06-25-10-34-14-11.png](https://user-images.githubusercontent.com/69170305/123371160-e9448e80-d5a0-11eb-8792-d51915e0a831.png)

Method to replicate:
1. Create a new gotify server (Or use one that has no apps setup).
2. Connect via a client android app. 
3. Go to Push Message tab
![Screenshot_2021-06-25-10-36-18-86.png](https://user-images.githubusercontent.com/69170305/123371315-3d4f7300-d5a1-11eb-8c8d-8bd311baa02a.png)
4. Then write down the title and content fields and press this button. (There is not app there)
![Screenshot_2021-06-25-10-36-30-70.png](https://user-images.githubusercontent.com/69170305/123371355-58ba7e00-d5a1-11eb-82f9-cfd765aaf668.png)
5. Freeze your android app. (Happens automatically 😊😊)

Here is whatever I could copy from the logs.

```
Saurav Adhikari, [25.06.21 14:35]
2021-06-25T08:49:29.917Z INFO: Entering LogsActivity
2021-06-25T08:49:27.397Z INFO: Loading more messages for -1
2021-06-25T08:49:27.335Z INFO: Entering MessagesActivity
2021-06-25T08:49:27.163Z INFO: Server version: 2.0.22@2021-04-16-18:16:19
2021-06-25T08:49:26.178Z INFO: WebSocket(1): opened
2021-06-25T08:49:25.541Z INFO: WebSocket(1): starting...
2021-06-25T08:49:25.523Z INFO: Starting WebSocketService
2021-06-25T08:49:25.516Z INFO: Create WebSocketService
2021-06-25T08:49:25.506Z INFO: Authenticated as admin
2021-06-25T08:49:23.272Z INFO: Entering InitializationActivity
2021-06-25T04:51:21.174Z INFO: Entering ShareActivity
2021-06-25T04:51:09.497Z INFO: Entering ShareActivity
2021-06-25T04:51:05.782Z INFO: Loading more messages for -1
2021-06-25T04:51:05.732Z INFO: Entering MessagesActivity
2021-06-25T04:51:05.620Z INFO: WebSocket(1): opened
2021-06-25T04:51:05.571Z INFO: Server version: 2.0.22@2021-04-16-18:16:19
2021-06-25T04:51:05.032Z INFO: WebSocket(1): starting...
2021-06-25T04:51:05.016Z INFO: Starting WebSocketService
2021-06-25T04:51:05.009Z INFO: Create WebSocketService
2021-06-25T04:51:05.001Z INFO: Authenticated as admin
2021-06-25T04:51:03.790Z INFO: Entering InitializationActivity
2021-06-25T04:49:06.521Z ERROR: uncaught exception
java.lang.ArrayIndexOutOfBoundsException: length=0; index=-1
 at java.util.ArrayList.get(ArrayList.java:439)
 at com.github.gotify.sharing.ShareActivity.pushMessage(ShareActivity.java:109)
 at com.github.gotify.sharing.ShareActivity_ViewBinding$1.doClick(ShareActivity_ViewBinding.java:40)
 at butterknife.internal.DebouncingOnClickListener.onClick(DebouncingOnClickListener.java:26)
 at android.view.View.performClick(View.java:6392)
 at com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1119)
 at android.view.View$PerformClick.run(View.java:25133)
 at android.os.Handler.handleCallback(Handler.java:790)
 at android.os.Handler.dispatchMessage(Handler.java:99)
 at android.os.Looper.loop(Looper.java:198)
 at android.app.ActivityThread.main(ActivityThread.java:7055)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:523)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:836)

2021-06-25T04:48:46.595Z INFO: Entering ShareActivity
2021-06-25T04:48:44.093Z INFO: Loading more messages for -1
2021-06-25T04:48:44.043Z INFO: Entering MessagesActivity
2021-06-25T04:48:43.893Z INFO: Server version: 2.0.22@2021-04-16-18:16:19
2021-06-25T04:48:43.877Z INFO: WebSocket(1): opened
2021-06-25T04:48:43.295Z INFO: WebSocket(1): starting...
2021-06-25T04:48:43.278Z INFO: Starting WebSocketService
2021-06-25T04:48:43.270Z INFO: Create WebSocketService
2021-06-25T04:48:43.261Z INFO: Authenticated as admin
2021-06-25T04:48:42.576Z INFO: Entering InitializationActivity
2021-06-25T04:45:54.417Z ERROR: uncaught exception
java.lang.ArrayIndexOutOfBoundsException: length=0; index=-1
 at java.util.ArrayList.get(ArrayList.java:439)
 at com.github.gotify.sharing.ShareActivity.pushMessage(ShareActivity.java:109)
 at com.github.gotify.sharing.ShareActivity_ViewBinding$1.doClick(ShareActivity_ViewBinding.java:40)
 at butterknife.internal.DebouncingOnClickListener.onClick(DebouncingOnClickListener.java:26)
 at android.view.View.performClick(View.java:6392)
 at com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1119)
 at android.view.View$PerformClick.run(View.java:25133)
 at android.os.Handler.handleCallback(Handler.java:790)
 at android.os.Handler.dispatchMessage(Handler.java:99)
 at android.os.Looper.loop(Looper.java:198)
 at android.app.ActivityThread.main(ActivityThread.java:7055)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:523)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:836)

Saurav Adhikari, [25.06.21 14:35]
2021-06-25T04:45:46.765Z INFO: Entering ShareActivity
2021-06-25T04:45:44.335Z INFO: Loading more messages for -1
2021-06-25T04:45:44.288Z INFO: Entering MessagesActivity
2021-06-25T04:45:44.173Z INFO: WebSocket(1): opened
2021-06-25T04:45:44.134Z INFO: Server version: 2.0.22@2021-04-16-18:16:19
2021-06-25T04:45:43.594Z INFO: WebSocket(1): starting...
2021-06-25T04:45:43.576Z INFO: Starting WebSocketService
2021-06-25T04:45:43.568Z INFO: Create WebSocketService
2021-06-25T04:45:43.558Z INFO: Authenticated as admin
2021-06-25T04:45:42.928Z INFO: Entering InitializationActivity
2021-06-25T04:45:27.289Z ERROR: uncaught exception
java.lang.ArrayIndexOutOfBoundsException: length=0; index=-1
 at java.util.ArrayList.get(ArrayList.java:439)
 at com.github.gotify.sharing.ShareActivity.pushMessage(ShareActivity.java:109)
 at com.github.gotify.sharing.ShareActivity_ViewBinding$1.doClick(ShareActivity_ViewBinding.java:40)
 at butterknife.internal.DebouncingOnClickListener.onClick(DebouncingOnClickListener.java:26)
 at android.view.View.performClick(View.java:6392)
 at com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1119)
 at android.view.View$PerformClick.run(View.java:25133)
 at android.os.Handler.handleCallback(Handler.java:790)
 at android.os.Handler.dispatchMessage(Handler.java:99)
 at android.os.Looper.loop(Looper.java:198)
 at android.app.ActivityThread.main(ActivityThread.java:7055)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:523)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:836)

2021-06-25T04:45:17.182Z INFO: Entering ShareActivity
2021-06-25T04:45:14.136Z INFO: Loading more messages for -1
2021-06-25T04:45:14.085Z INFO: Entering MessagesActivity
2021-06-25T04:45:14.034Z INFO: WebSocket(1): opened
2021-06-25T04:45:13.956Z INFO: Server version: 2.0.22@2021-04-16-18:16:19
2021-06-25T04:45:13.427Z INFO: WebSocket(1): starting...
2021-06-25T04:45:13.403Z INFO: Starting WebSocketService
2021-06-25T04:45:13.389Z INFO: Create WebSocketService
2021-06-25T04:45:13.370Z INFO: Authenticated as admin
2021-06-25T04:45:12.736Z INFO: Entering InitializationActivity
2021-06-25T04:44:40.861Z INFO: Entering LoginActivity
2021-06-25T04:44:39.420Z ERROR: Error while api call
Code(0) Response: 
 at com.github.gotify.api.Callback$RetrofitCallback.onFailure(Callback.java:80)
 at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$2.run(ExecutorCallAdapterFactory.java:80)
 at android.os.Handler.handleCallback(Handler.java:790)
 at android.os.Handler.dispatchMessage(Handler.java:99)
 at android.os.Looper.loop(Looper.java:198)
 at android.app.ActivityThread.main(ActivityThread.java:7055)
 at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:523)
 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:836)
Caused by: javax.net.ssl.SSLHandshakeException: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found.
 at com.android.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:219)
 at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.java:319)
 at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.java:283)
 at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:168)
 at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
 at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
 at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
 at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)

Saurav Adhikari, [25.06.21 14:35]
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
 at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
 at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
 at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
 at com.github.gotify.client.auth.ApiKeyAuth.intercept(ApiKeyAuth.java:66)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
 at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
 at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254)
 at okhttp3.RealCall$AsyncCall.execute(RealCall.java:200)
 at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
 at java.lang.Thread.run(Thread.java:764)
Caused by: java.security.cert.CertificateException: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found.
 at com.android.org.conscrypt.TrustManagerImpl.verifyChain(TrustManagerImpl.java:661)
 at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:539)
 at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:495)
 at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:418)
 at com.android.org.conscrypt.TrustManagerImpl.getTrustedChainForServer(TrustManagerImpl.java:339)
 at android.security.net.config.NetworkSecurityTrustManager.checkServerTrusted(NetworkSecurityTrustManager.java:94)
 at android.security.net.config.RootTrustManager.checkServerTrusted(RootTrustManager.java:88)
 at com.android.org.conscrypt.Platform.checkServerTrusted(Platform.java:197)
 at com.android.org.conscrypt.ConscryptFileDescriptorSocket.verifyCertificateChain(ConscryptFileDescriptorSocket.java:399)
 at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
 at com.android.org.conscrypt.SslWrapper.doHandshake(SslWrapper.java:374)
 at com.android.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:217)
 ... 26 more
Caused by: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found.
```

--- a/app/src/main/java/com/github/gotify/messages/provider/ApplicationHolder.java
+++ b/app/src/main/java/com/github/gotify/messages/provider/ApplicationHolder.java
@@ -36,14 +36,12 @@ public class ApplicationHolder {
 
     private void onReceiveApps(List<Application> apps) {
         state = apps;
-        if (onUpdate != null)
-            onUpdate.run();
+        if (onUpdate != null) onUpdate.run();
     }
 
     private void onFailedApps(ApiException e) {
         Utils.showSnackBar(activity, "Could not request applications, see logs.");
-        if (onUpdateFailed != null)
-            onUpdateFailed.run();
+        if (onUpdateFailed != null) onUpdateFailed.run();
     }
 
     public List<Application> get() {

--- a/app/src/main/java/com/github/gotify/messages/provider/ApplicationHolder.java
+++ b/app/src/main/java/com/github/gotify/messages/provider/ApplicationHolder.java
@@ -13,6 +13,7 @@ import java.util.List;
 public class ApplicationHolder {
     private List<Application> state;
     private Runnable onUpdate;
+    private Runnable onUpdateFailed;
     private Activity activity;
     private ApiClient client;
 
@@ -35,11 +36,14 @@ public class ApplicationHolder {
 
     private void onReceiveApps(List<Application> apps) {
         state = apps;
-        onUpdate.run();
+        if (onUpdate != null)
+            onUpdate.run();
     }
 
     private void onFailedApps(ApiException e) {
         Utils.showSnackBar(activity, "Could not request applications, see logs.");
+        if (onUpdateFailed != null)
+            onUpdateFailed.run();
     }
 
     public List<Application> get() {
@@ -48,5 +52,9 @@ public class ApplicationHolder {
 
     public void onUpdate(Runnable runnable) {
         this.onUpdate = runnable;
+    }
+
+    public void onUpdateFailed(Runnable runnable) {
+        this.onUpdateFailed = runnable;
     }
 }

--- a/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
+++ b/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
@@ -81,14 +81,15 @@ public class ShareActivity extends AppCompatActivity {
         ApiClient client =
                 ClientFactory.clientToken(settings.url(), settings.sslSettings(), settings.token());
         appsHolder = new ApplicationHolder(this, client);
-        appsHolder.onUpdate(() -> {
-            List<Application> apps = appsHolder.get();
-            populateSpinner(apps);
+        appsHolder.onUpdate(
+                () -> {
+                    List<Application> apps = appsHolder.get();
+                    populateSpinner(apps);
 
-            boolean appsAvailable = !apps.isEmpty();
-            pushMessageButton.setEnabled(appsAvailable);
-            missingAppsInfo.setVisibility(appsAvailable ? View.GONE : View.VISIBLE);
-        });
+                    boolean appsAvailable = !apps.isEmpty();
+                    pushMessageButton.setEnabled(appsAvailable);
+                    missingAppsInfo.setVisibility(appsAvailable ? View.GONE : View.VISIBLE);
+                });
         appsHolder.onUpdateFailed(() -> pushMessageButton.setEnabled(false));
         appsHolder.request();
     }
@@ -115,8 +116,8 @@ public class ShareActivity extends AppCompatActivity {
             Toast.makeText(this, "Priority should be number.", Toast.LENGTH_LONG).show();
             return;
         } else if (appIndex == Spinner.INVALID_POSITION) {
-            //For safety, e.g. loading the apps needs too much time (maybe a timeout) and
-            //the user tries to push without an app selected.
+            // For safety, e.g. loading the apps needs too much time (maybe a timeout) and
+            // the user tries to push without an app selected.
             Toast.makeText(this, "An app must be selected.", Toast.LENGTH_LONG).show();
             return;
         }

--- a/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
+++ b/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
@@ -6,6 +6,7 @@ import android.os.Bundle;
 import android.view.MenuItem;
 import android.view.View;
 import android.widget.ArrayAdapter;
+import android.widget.Button;
 import android.widget.EditText;
 import android.widget.Spinner;
 import android.widget.Toast;
@@ -46,6 +47,9 @@ public class ShareActivity extends AppCompatActivity {
     @BindView(R.id.appSpinner)
     Spinner appSpinner;
 
+    @BindView(R.id.push_button)
+    Button pushMessageButton;
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
@@ -73,7 +77,11 @@ public class ShareActivity extends AppCompatActivity {
         ApiClient client =
                 ClientFactory.clientToken(settings.url(), settings.sslSettings(), settings.token());
         appsHolder = new ApplicationHolder(this, client);
-        appsHolder.onUpdate(() -> populateSpinner(appsHolder.get()));
+        appsHolder.onUpdate(() -> {
+            List<Application> apps = appsHolder.get();
+            populateSpinner(apps);
+            pushMessageButton.setEnabled(apps.size() > 0);
+        });
         appsHolder.request();
     }
 

--- a/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
+++ b/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
@@ -85,10 +85,11 @@ public class ShareActivity extends AppCompatActivity {
             List<Application> apps = appsHolder.get();
             populateSpinner(apps);
 
-            boolean appsAvailable = apps.size() > 0;
+            boolean appsAvailable = !apps.isEmpty();
             pushMessageButton.setEnabled(appsAvailable);
             missingAppsInfo.setVisibility(appsAvailable ? View.GONE : View.VISIBLE);
         });
+        appsHolder.onUpdateFailed(() -> pushMessageButton.setEnabled(false));
         appsHolder.request();
     }
 
@@ -112,6 +113,11 @@ public class ShareActivity extends AppCompatActivity {
             return;
         } else if (priority.isEmpty()) {
             Toast.makeText(this, "Priority should be number.", Toast.LENGTH_LONG).show();
+            return;
+        } else if (appIndex == Spinner.INVALID_POSITION) {
+            //For safety, e.g. loading the apps needs too much time (maybe a timeout) and
+            //the user tries to push without an app selected.
+            Toast.makeText(this, "An app must be selected.", Toast.LENGTH_LONG).show();
             return;
         }
 

--- a/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
+++ b/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
@@ -9,6 +9,7 @@ import android.widget.ArrayAdapter;
 import android.widget.Button;
 import android.widget.EditText;
 import android.widget.Spinner;
+import android.widget.TextView;
 import android.widget.Toast;
 import androidx.appcompat.app.ActionBar;
 import androidx.appcompat.app.AppCompatActivity;
@@ -50,6 +51,9 @@ public class ShareActivity extends AppCompatActivity {
     @BindView(R.id.push_button)
     Button pushMessageButton;
 
+    @BindView(R.id.infoMissingApps)
+    TextView missingAppsInfo;
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
@@ -80,9 +84,14 @@ public class ShareActivity extends AppCompatActivity {
         appsHolder.onUpdate(() -> {
             List<Application> apps = appsHolder.get();
             populateSpinner(apps);
-            pushMessageButton.setEnabled(apps.size() > 0);
+            preventPushing(apps.size() == 0);
         });
         appsHolder.request();
+    }
+
+    private void preventPushing(boolean prevent) {
+        pushMessageButton.setEnabled(!prevent);
+        missingAppsInfo.setVisibility(prevent ? View.VISIBLE : View.GONE);
     }
 
     @Override

--- a/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
+++ b/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
@@ -84,14 +84,12 @@ public class ShareActivity extends AppCompatActivity {
         appsHolder.onUpdate(() -> {
             List<Application> apps = appsHolder.get();
             populateSpinner(apps);
-            preventPushing(apps.size() == 0);
+
+            boolean appsAvailable = apps.size() > 0;
+            pushMessageButton.setEnabled(appsAvailable);
+            missingAppsInfo.setVisibility(appsAvailable ? View.GONE : View.VISIBLE);
         });
         appsHolder.request();
-    }
-
-    private void preventPushing(boolean prevent) {
-        pushMessageButton.setEnabled(!prevent);
-        missingAppsInfo.setVisibility(prevent ? View.VISIBLE : View.GONE);
     }
 
     @Override

--- a/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
+++ b/app/src/main/java/com/github/gotify/sharing/ShareActivity.java
@@ -8,8 +8,8 @@ import android.view.View;
 import android.widget.ArrayAdapter;
 import android.widget.Button;
 import android.widget.EditText;
+import android.widget.LinearLayout;
 import android.widget.Spinner;
-import android.widget.TextView;
 import android.widget.Toast;
 import androidx.appcompat.app.ActionBar;
 import androidx.appcompat.app.AppCompatActivity;
@@ -51,8 +51,8 @@ public class ShareActivity extends AppCompatActivity {
     @BindView(R.id.push_button)
     Button pushMessageButton;
 
-    @BindView(R.id.infoMissingApps)
-    TextView missingAppsInfo;
+    @BindView(R.id.missingAppsContainer)
+    LinearLayout missingAppsInfo;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {

--- a/app/src/main/res/drawable/ic_info.xml
+++ b/app/src/main/res/drawable/ic_info.xml
@@ -1,0 +1,9 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24">
+  <path
+      android:fillColor="@color/icons"
+      android:pathData="M12,2C6.48,2 2,6.48 2,12s4.48,10 10,10 10,-4.48 10,-10S17.52,2 12,2zM13,17h-2v-6h2v6zM13,9h-2L11,7h2v2z"/>
+</vector>

--- a/app/src/main/res/layout/activity_share.xml
+++ b/app/src/main/res/layout/activity_share.xml
@@ -1,104 +1,109 @@
 <?xml version="1.0" encoding="utf-8"?>
-<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<ScrollView
+    xmlns:android="http://schemas.android.com/apk/res/android"
     android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    android:orientation="vertical">
-
-    <include
-        layout="@layout/app_bar_drawer"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content" />
-
-    <com.google.android.material.textfield.TextInputLayout
-        style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_margin="20dp">
-
-        <EditText
-            android:id="@+id/title"
-            android:layout_width="match_parent"
-            android:layout_height="wrap_content"
-            android:ems="10"
-            android:hint="@string/push_title_hint"
-            android:inputType="text"
-            android:singleLine="true" />
-    </com.google.android.material.textfield.TextInputLayout>
-
-    <com.google.android.material.textfield.TextInputLayout
-        style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_marginHorizontal="20dp"
-        android:layout_marginBottom="20dp">
-
-        <com.google.android.material.textfield.TextInputEditText
-            android:id="@+id/content"
-            android:layout_width="match_parent"
-            android:layout_height="wrap_content"
-            android:hint="@string/push_content_hint"
-            android:inputType="textCapSentences|textMultiLine"
-            android:maxLength="2000"
-            android:maxLines="10" />
-    </com.google.android.material.textfield.TextInputLayout>
-
-    <com.google.android.material.textfield.TextInputLayout
-        style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_marginHorizontal="20dp"
-        android:layout_marginBottom="20dp">
-
-        <com.google.android.material.textfield.TextInputEditText
-            android:id="@+id/edtTxtPriority"
-            android:layout_width="match_parent"
-            android:layout_height="wrap_content"
-            android:digits="0123456789"
-            android:hint="@string/push_priority_hint"
-            android:imeOptions="actionDone"
-            android:inputType="numberSigned"
-            android:maxLength="3"
-            android:text="0" />
-    </com.google.android.material.textfield.TextInputLayout>
+    android:layout_height="match_parent">
 
     <LinearLayout
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
-        android:layout_marginHorizontal="20dp"
-        android:layout_marginBottom="20dp">
+        android:orientation="vertical">
+
+        <include
+            layout="@layout/app_bar_drawer"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content" />
+
+        <com.google.android.material.textfield.TextInputLayout
+            style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_margin="20dp">
+
+            <EditText
+                android:id="@+id/title"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:ems="10"
+                android:hint="@string/push_title_hint"
+                android:inputType="text"
+                android:singleLine="true" />
+        </com.google.android.material.textfield.TextInputLayout>
+
+        <com.google.android.material.textfield.TextInputLayout
+            style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginHorizontal="20dp"
+            android:layout_marginBottom="20dp">
+
+            <com.google.android.material.textfield.TextInputEditText
+                android:id="@+id/content"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:hint="@string/push_content_hint"
+                android:inputType="textCapSentences|textMultiLine"
+                android:maxLength="2000"
+                android:maxLines="10" />
+        </com.google.android.material.textfield.TextInputLayout>
+
+        <com.google.android.material.textfield.TextInputLayout
+            style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginHorizontal="20dp"
+            android:layout_marginBottom="20dp">
+
+            <com.google.android.material.textfield.TextInputEditText
+                android:id="@+id/edtTxtPriority"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:digits="0123456789"
+                android:hint="@string/push_priority_hint"
+                android:imeOptions="actionDone"
+                android:inputType="numberSigned"
+                android:maxLength="3"
+                android:text="0" />
+        </com.google.android.material.textfield.TextInputLayout>
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginHorizontal="20dp"
+            android:layout_marginBottom="20dp">
+
+            <TextView
+                android:id="@+id/txtAppListDesc"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:text="@string/appListDescription" />
+
+            <Spinner
+                android:id="@+id/appSpinner"
+                android:layout_width="wrap_content"
+                android:layout_height="50dp"
+                android:spinnerMode="dropdown"
+                android:textSize="18sp" />
+        </LinearLayout>
 
         <TextView
-            android:id="@+id/txtAppListDesc"
-            android:layout_width="wrap_content"
+            android:id="@+id/infoMissingApps"
+            android:layout_width="match_parent"
             android:layout_height="wrap_content"
-            android:text="@string/appListDescription" />
+            android:layout_marginHorizontal="20dp"
+            android:layout_marginBottom="20dp"
+            android:text="@string/push_missing_app_info"
+            android:textAlignment="center"
+            android:visibility="gone" />
 
-        <Spinner
-            android:id="@+id/appSpinner"
-            android:layout_width="wrap_content"
-            android:layout_height="50dp"
-            android:spinnerMode="dropdown"
-            android:textSize="18sp" />
+        <Button
+            android:id="@+id/push_button"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center_vertical|center_horizontal"
+            android:layout_marginBottom="20dp"
+            android:layout_marginHorizontal="20dp"
+            android:text="@string/push_button" />
+
     </LinearLayout>
-
-    <TextView
-        android:id="@+id/infoMissingApps"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_marginHorizontal="20dp"
-        android:layout_marginBottom="20dp"
-        android:text="@string/push_missing_app_info"
-        android:textAlignment="center"
-        android:textColor="#BDBDBD"
-        android:visibility="gone" />
-
-    <Button
-        android:id="@+id/push_button"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_gravity="center_vertical|center_horizontal"
-        android:layout_marginBottom="20dp"
-        android:layout_marginHorizontal="20dp"
-        android:text="@string/push_button" />
-
-</LinearLayout>
+</ScrollView>

--- a/app/src/main/res/layout/activity_share.xml
+++ b/app/src/main/res/layout/activity_share.xml
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <ScrollView
     xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
     android:layout_width="match_parent"
     android:layout_height="match_parent">
 
@@ -86,15 +87,39 @@
                 android:textSize="18sp" />
         </LinearLayout>
 
-        <TextView
-            android:id="@+id/infoMissingApps"
+        <LinearLayout
+            android:id="@+id/missingAppsContainer"
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:layout_marginHorizontal="20dp"
             android:layout_marginBottom="20dp"
-            android:text="@string/push_missing_app_info"
-            android:textAlignment="center"
-            android:visibility="gone" />
+            android:gravity="center"
+            android:orientation="horizontal"
+            android:paddingStart="10dp"
+            android:paddingEnd="10dp"
+            android:visibility="gone">
+
+            <ImageView
+                android:id="@+id/missingAppsIcon"
+                android:layout_width="40dp"
+                android:layout_height="match_parent"
+                android:layout_weight="0"
+                app:srcCompat="@drawable/ic_info" />
+
+            <Space
+                android:layout_width="10dp"
+                android:layout_height="match_parent"
+                android:layout_weight="0" />
+
+            <TextView
+                android:id="@+id/missingAppsText"
+                android:layout_width="wrap_content"
+                android:layout_height="match_parent"
+                android:layout_weight="0"
+                android:text="@string/push_missing_app_info"
+                android:textAlignment="center"
+                android:visibility="visible" />
+        </LinearLayout>
 
         <Button
             android:id="@+id/push_button"

--- a/app/src/main/res/layout/activity_share.xml
+++ b/app/src/main/res/layout/activity_share.xml
@@ -81,6 +81,17 @@
             android:textSize="18sp" />
     </LinearLayout>
 
+    <TextView
+        android:id="@+id/infoMissingApps"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginHorizontal="20dp"
+        android:layout_marginBottom="20dp"
+        android:text="@string/push_missing_app_info"
+        android:textAlignment="center"
+        android:textColor="#BDBDBD"
+        android:visibility="gone" />
+
     <Button
         android:id="@+id/push_button"
         android:layout_width="match_parent"
@@ -89,4 +100,5 @@
         android:layout_marginBottom="20dp"
         android:layout_marginHorizontal="20dp"
         android:text="@string/push_button" />
+
 </LinearLayout>

--- a/app/src/main/res/layout/activity_share.xml
+++ b/app/src/main/res/layout/activity_share.xml
@@ -101,8 +101,9 @@
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:layout_gravity="center_vertical|center_horizontal"
-            android:layout_marginBottom="20dp"
             android:layout_marginHorizontal="20dp"
+            android:layout_marginBottom="20dp"
+            android:enabled="false"
             android:text="@string/push_button" />
 
     </LinearLayout>

--- a/app/src/main/res/layout/activity_share.xml
+++ b/app/src/main/res/layout/activity_share.xml
@@ -128,7 +128,6 @@
             android:layout_gravity="center_vertical|center_horizontal"
             android:layout_marginHorizontal="20dp"
             android:layout_marginBottom="20dp"
-            android:enabled="false"
             android:text="@string/push_button" />
 
     </LinearLayout>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -81,6 +81,6 @@
     <string name="push_title_hint">Title</string>
     <string name="push_content_hint">Content</string>
     <string name="push_priority_hint">Priority</string>
-    <string name="push_missing_app_info">There are no applications available on the server\nto push a message to.</string>
+    <string name="push_missing_app_info">There are no applications available on the server to push a message to.</string>
     <string name="message_copied_to_clipboard">Content copied to clipboard</string>
 </resources>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -81,5 +81,6 @@
     <string name="push_title_hint">Title</string>
     <string name="push_content_hint">Content</string>
     <string name="push_priority_hint">Priority</string>
+    <string name="push_missing_app_info">There are no applications available on the server\nto push a message to.</string>
     <string name="message_copied_to_clipboard">Content copied to clipboard</string>
 </resources>

